src/PhysicsEngine.cpp

```diff
@@ -0,0 +1,119 @@
+#include "PhysicsEngine.h"
+#include "Entity.h"
+#include "AK/AK.h"
+
+void PhysicsEngine::initialize()
+{
+    auto const physics_engine = std::make_shared<PhysicsEngine>();
+    set_instance(physics_engine);
+}
+
+void PhysicsEngine::update_physics()
+{
+    solve_collisions();
+}
+
+void PhysicsEngine::emplace_collider(std::shared_ptr<Collider2D> const& collider)
+{
+    colliders.emplace_back(collider);
+}
+
+void PhysicsEngine::remove_collider(std::shared_ptr<Collider2D> const& collider)
+{
+    // Check if the shared pointer is already in the vector
+    auto const it = std::find_if(colliders.begin(), colliders.end(),
+        [&](const std::weak_ptr<Collider2D>& wp) {
+            return wp.lock() == collider;
+        });
+
+    if (it != colliders.end())
+        colliders.erase(it);
+}
+
+void PhysicsEngine::solve_collisions()
+{
+    // Collision detection
+    for (u32 i = 0; i < colliders.size(); i++)
+    {
+        for (u32 j = 0; j < colliders.size(); j++)
+        {
+            if (i == j)
+                continue;
+
+            ColliderType2D const col1 = colliders[i].lock()->get_collider_type();
+            ColliderType2D const col2 = colliders[j].lock()->get_collider_type();
+            std::shared_ptr<Collider2D> obj1 = colliders[i].lock();
+            std::shared_ptr<Collider2D> obj2 = colliders[j].lock();
+
+            // CIRCLE/CIRCLE
+            if (col1 == col2 && col1 == ColliderType2D::Circle)
+            {
+                const glm::vec3 posi = colliders[i].lock()->entity->transform->get_local_position();
+                const glm::vec3 posj = colliders[j].lock()->entity->transform->get_local_position();
+
+                const glm::vec2 posi_2d = AK::convert_3d_to_2d(posi);
+                const glm::vec2 posj_2d = AK::convert_3d_to_2d(posj);
+
+                const float posDistance = glm::distance(posi_2d, posj_2d);
+                const float radiusSum = colliders[i].lock()->get_radius_2d() + colliders[j].lock()->get_radius_2d();
+
+                if (posDistance < radiusSum)
+                {
+                    obj1->separate(posi, posj, obj1->get_radius_2d(), obj2->get_radius_2d());
+                    obj2->separate(posj, posi, obj2->get_radius_2d(), obj1->get_radius_2d());
+                }
```

Comment:
There are lots of locks here, we should just lock once, at the beginning of the context

Also, lots of west const here

Also, camelCase in posDistance and radiusSum